rts/motoko-rts/src/region.rs

```diff
@@ -0,0 +1,854 @@
+        pub fn get(r: &RegionId) -> Option<RegionSizeInPages> {
+            RegionSizeInPages::from_u64(read_u64(index(r)))
+        }
+
+        pub fn set(r: &RegionId, s: Option<RegionSizeInPages>) {
+            write_u64(index(r), RegionSizeInPages::into_u64(s))
+        }
+    }
+}
+
+unsafe fn alloc_region<M: Memory>(
+    mem: &mut M,
+    id: u16,
+    page_count: u32,
+    vec_pages: Value,
+) -> Value {
+    let r_ptr = mem.alloc_words(size_of::<Region>());
+
+    // NB. cannot use as_region() here as we didn't write the header yet
+    let region = r_ptr.get_ptr() as *mut Region;
+    (*region).header.tag = TAG_REGION;
+    (*region).header.init_forward(r_ptr);
+    (*region).id = id;
+    // The padding must be initialized with zero because it is read by the compiler-generated code.
+    (*region).zero_padding = 0;
+    (*region).page_count = page_count;
+    init_with_barrier(mem, &mut (*region).vec_pages, vec_pages);
+
+    r_ptr
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_id<M: Memory>(_mem: &mut M, r: Value) -> u32 {
+    let r = r.as_region();
+    (*r).id.into()
+}
+
+// Region manager's total memory size in stable memory, in _pages_.
+#[ic_mem_fn]
+pub unsafe fn region_get_mem_size<M: Memory>(_mem: &mut M) -> u64 {
+    let size = {
+        if let Some(s) = crate::region::REGION_SET_MEM_SIZE {
+            return s;
+        };
+        if crate::region::REGION_MEM_SIZE_INIT {
+            meta_data::size::STATIC_MEM_IN_PAGES as u64
+                + crate::region::REGION_TOTAL_ALLOCATED_BLOCKS as u64
+                    * (meta_data::size::PAGES_IN_BLOCK as u64)
+        } else {
+            // Before initialization of anything, give back zero.
+            0
+        }
+    };
+    size
+}
+
+// Region manager's total memory size in stable memory, in _pages_.
+#[ic_mem_fn]
+pub unsafe fn region_set_mem_size<M: Memory>(_mem: &mut M, size: u64) {
+    crate::region::REGION_SET_MEM_SIZE = Some(size);
+}
+
+// Helper for commmon logic that reserves low-valued RegionIds in a certain span for future use.
+// When first is some, we are actually reserving.  When first is none, we are checking that the reservation has occured.
+unsafe fn region_reserve_id_span<M: Memory>(_mem: &mut M, first: Option<RegionId>, last: RegionId) {
+    if let Some(first) = first {
+        let next_id = meta_data::total_allocated_regions::get() as u16;
+        assert_eq!(first.0, next_id);
+        assert!(first.0 <= last.0);
+        meta_data::total_allocated_regions::set((last.0 + 1) as u64);
+    } else {
+        let next_id = meta_data::total_allocated_regions::get() as u16;
+        assert!(last.0 < next_id);
+    }
+}
+
+#[ic_mem_fn]
+pub unsafe fn region_new<M: Memory>(mem: &mut M) -> Value {
+    let next_id = meta_data::total_allocated_regions::get() as u16;
+    meta_data::total_allocated_regions::set(next_id as u64 + 1);
+
+    let vec_pages = alloc_blob(mem, Bytes(0));
+    let r_ptr = alloc_region(mem, next_id, 0, vec_pages);
```

Comment:
Required allocation barrier for the incremental GC:

```suggestion
    allocation_barrier(vec_pages);
    let r_ptr = alloc_region(mem, next_id, 0, vec_pages);
```

Reply:
`vec_pages` is a Blob -- so is this still required? (looking at doc for `allocation_barrier` suggests that it is not required for `Blob`s`)